Add implementation status doc (#130)
BerkeleyLab/caffeine: docs/implementation-status.md

# Implementation Status

Caffeine is an implementation of the Parallel Runtime Interface for Fortran (PRIF). This document
outlines the implementation status in Caffeine of the features defined in the

> [latest PRIF revision](https://dx.doi.org/10.25344/S4WG64). Caffeine contains interfaces for all

Reviewer: ```suggestion
[latest PRIF specification, revision 0.4](https://dx.doi.org/10.25344/S4WG64). Caffeine contains interfaces for all
```

> of the PRIF procedures and are callable, but some procedure definitions are unimplemented. For

Reviewer: ```suggestion
of the PRIF procedures and the symbols are linkable and callable, but some procedures will fail at runtime with an unimplemented error. For
```

more details about the implementation of the various PRIF features, please see the
following sections:

- [Named Constants](#Named-Constants)
- [`stat` and `errmsg` support](#stat-and-errmsg-support) - NO support yet
- [Program Startup and Shutdown](#Program-Startup-and-Shutdown)
- [Image Queries](#Image-Queries)
- [Storage Management](#Storage-Management)
- [Coarray Queries](#Coarray-Queries)
- [Continguous Coarray Access](#Continguous-Coarray-Access)
- [Strided Coarray Access](#Strided-Coarray-Access)
- [SYNC Statements](#SYNC-Statements)
- [Locks and Unlocks](#Locks-and-Unlocks)
- [Critical](#Critical)
- [Events and Notifications](#Events-and-Notifications)
- [Teams](#teams)
- [Collectives](#Collectives)
- [Atomic Memory Operations](#Atomic-Memory-Operations)

> TODO: For features with all no support, flesh out table, but comment it out

Reviewer: ```suggestion
```

## Named Constants

Caffeine contains definitions for all of the PRIF-relevant constants from ISO_FORTRAN_ENV and for
all of the PRIF-specific constants.

## `stat` and `errmsg` support

Many PRIF procedures have optional arguments `stat`, `errmsg`, and `errmsg_alloc`. There is no
support for these optional arguments in Caffeine yet and this is not represented in the tables
in the following sections.

## Program Startup and Shutdown
### Support = partial

| Procedure | Status | Notes |
|-----------|--------|-------|

> | `prif_init` | yes | n/a |

Reviewer: Looking over the rendered doc, the repeated `n/a` in the Notes column is somewhat visually distracting and makes the tables look "busier" than they are. I suggest we just leave the Notes cell blank when we have nothing to say.

Author: Great point. I have removed the `n/a` notes from the tables.

| `prif_stop`, `prif_error_stop` | partial | Missing support for `quiet=.true.` |
| `prif_fail_image` | no | n/a |

---

## Image Queries
### Support = partial

| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_num_images`    | yes | n\a |
| `prif_num_images_with_team`, `prif_num_images_with_team_number`  | no | n\a |

> | `prif_this_image_no_coarray` | yes | n\a |

Reviewer: ```suggestion
| `prif_this_image_no_coarray` | partial | team argument is ignored |
```

| `prif_this_image_with_coarray`, `prif_this_image_with_dim`  | no | n\a |
| `prif_failed_images` | no | n/a |
| `prif_stopped_images` | no | n/a |
| `prif_image_status` | no | n/a |

---

## Storage Management
### Support = partial (no support for coarray aliases)

| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_allocate_coarray`    | yes | n/a |
| `prif_allocate`    | yes | n/a |

> | `prif_deallocate_coarray`    | partial | no `final_func` arg support |
> | `prif_deallocate`    | yes | n/a |
> | `prif_alias_create`    | no | n/a |

Reviewer: On a cosmetic note, even in the absence of color we could potentially still make the Status column "pop" a bit more, eg:

| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_allocate_coarray`    | **YES** |  |
| `prif_allocate`    | **YES** |  |
| `prif_deallocate_coarray`    | *partial* | no `final_func` arg support |
| `prif_deallocate`    | **YES** |  |
| `prif_alias_create`    | no |  |
| `prif_alias_destroy`    | no | |

However if this seems like too much hassle to maintain then we could keep the current convention.

```suggestion
| `prif_deallocate_coarray`    | *partial* | no `final_func` arg support |
| `prif_deallocate`    | **YES** | n/a |
| `prif_alias_create`    | no | n/a |
```

<!---
| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_lock`            | no  | n\a |
| `prif_lock_indirect`   | no  | n\a |
| `prif_unlock`          | no  | n\a |
| `prif_unlock_indirect` | no  | n\a |
-->

---

## Critical
### Support = no

<!---
| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_critical`     | no  | n\a |
| `prif_end_critical` | no  | n\a |
-->

---

## Events and Notifications
### Support = no

<!---
| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_event_post`          | no  | n\a |
| `prif_event_post_indirect` | no  | n\a |
| `prif_event_wait`          | no  | n\a |
| `prif_event_query`         | no  | n\a |
| `prif_notify_wait`         | no  | n\a |
-->

---

## Teams
### Support = partial (No support for `prif_get_team` and `prif_team_number`)

| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_form_team`   | yes | n\a |
| `prif_get_team`    | no  | n\a |
| `prif_team_number` | no  | n\a |
| `prif_change_team` | yes | n\a |
| `prif_end_team`    | yes | n\a |

---

## Collectives
### Support = partial (...)

| Procedure | Status | Notes |
|-----------|--------|-------|

> | `prif_co_broadcast` | ... | ... |
> | `prif_co_max`       | ... | ... |
> | `prif_co_min`       | ... | ... |
> | `prif_co_sum`       | ... | ... |
> | `prif_co_reduce`    | ... | ... |

Reviewer: ```suggestion
| `prif_co_broadcast` | partial | no support for derived types with `allocatable` components |
| `prif_co_max`       | partial | only supports 32-bit and 64-bit numeric types |
| `prif_co_min`       | partial | only supports 32-bit and 64-bit numeric types |
| `prif_co_sum`       | partial | only supports 32-bit and 64-bit numeric types |
| `prif_co_reduce`    | partial | only supports intrinsic types (no support for derived types) |
```

Reviewer: `prif_co_{min,max}` have an implementation for character strings, but @rouson and I have just discovered that it's "fragile" (issue #132) such that we might not want to claim support

| Procedure | Status | Notes |
|-----------|--------|-------|
| `prif_co_broadcast` | *partial* | no support for derived types with `allocatable` components |
| `prif_co_max`       | *partial* | only supports 32-bit and 64-bit numeric types |

> | `prif_co_min`       | no        |  |
> | `prif_co_sum`       | no        |  |

Reviewer: Why were these two changed from the suggestion? 

I believe all three built-in reduction operations (`co_{min,max,sum}`) are currently in the same state, ie they are believed to be working with gfortran and 32/64-bit numeric datatypes. 

These all have some known problems (e.g. issue #132, which is really a PRIF spec issue).  `co_{min,max}` have some support for Fortran strings but it currently relies on undefined behavior of gfortran. None of these work on flang yet, but we're not publicly claiming flang support for anything yet.

However despite all that I think I'd still characterize all three the same way:

```suggestion
| `prif_co_min`       | *partial* | only supports 32-bit and 64-bit numeric types |
| `prif_co_sum`       | *partial* | only supports 32-bit and 64-bit numeric types |
```

Reviewer: Re-reading the prior comment thread I think I understand the miscommunication. 

The concerns with `prif_co_{min,max}` are almost entirely regarding the implementation of character type arguments. IMO 32/64-bit numeric datatypes should be considered working.

Author: @bonachea Thanks, yes I just misunderstood your earlier comment. Thank you for catching that are for the code suggestion.
